src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h

```diff
@@ -47,6 +48,12 @@ namespace theia {
 // Computes the global rotations given relative rotations and an initial guess
 // for the global orientations. Nonlinear optimization is performed with Ceres
 // using a SoftL1 loss function to be robust to outliers.
+//
+// CostFunctionGenerator is a class which defines a function `Create` that
+// generates a ceres::CostFunction* given a relative rotation in vector 3D:
+// ceres::CostFunction* Create(const Eigen::Vector3d& relative_rotation)
+// Check PairwiseRotationError as an example.
+template <class CostFunctionGenerator=PairwiseRotationError>
```

Comment:
I think in general this does what we want. Template classes in C++ are a bit hard to wrangle though, e.g. we have to do what you did here which is putting everying in the header and that adds to code size and compile speed. I proposed another soln which uses `std::function`, feel free to take a look and decide which feels more natural.

Reply:
Thanks for the suggestion. I had a look at your changes and it looks good to me except the fact that the with `std::function`, it's not flexible to have different parameters for different cost functions, which happens very likely.

It's actually possible to separate the template into header and source file, but in source file it looks quite ugly, and I think we also need to specify the templates explicitly in source file.